### DISCUSSION TITLE
Possible to Generate schema.graphql File on Project Build?

### DISCUSSION THREAD

**Discussion by Presipope** [graphql-dotnet/graphql-dotnet#3604] [Q&A] [answered]
I'm wondering if there's a way to use some combination of the schema builder, schema printer, or other classes to generate a schema.graphql file every time my project using graphql builds. This would be used for the client side application that is using relay. 

My idea is to create a that does the building of the schema and prints it to a file. Then create a custom build target in my csproj file so this runs every time. 

I'm thinking something along the lines of: 
```c#
public static class SchemaGenerator
{
   public static void BuildSchema()
   {
      // code in here to generate the schema based on my existing C# structure, similar to how the app does this at start up to begin with
   }
}

public class Program
{
    public static void Main(string[] args)
    {
        SchemaGenerator.BuildSchema();
    }
}
``` 

Then add this as a custom build target in the csproj file. 

Also open to any other ideas. I've attempted doing an introspection query on the build side of the client app, but that seems a little backwards. Not to mention I'll run into a lot of weird authorization issues I'm sure with our CI/CD pipeline and local IIS servers during dev. 

I've been attempting to fumble my way through online documentation and the repository but I'm a little too new to this level of deep diving in repos on GitHub to fully grasp my way through this problem.

**Comment by Shane32** [comment]
You’re on the right track. I’m not sure that there is any specific advice here. Let me know if you have a specific question as you attempt to implement it.

**Reply by Presipope** [reply]
Just posted what I got working, I'd be interested in hearing if you could see any glaring pitfalls with what I ended up with. Thanks for making yourself available in case I had had any questions!

**Comment by sungam3r** [comment]
> I'm wondering if there's a way 

There is. Just describe your case, your inputs. For example, you may spin-up https://github.com/sungam3r/graphql-sdl-exporter to generate SDL from URL of your app.

**Reply by Presipope** [reply]
Saw this after I posted my response lol. 

I like that the tool can do a url or executable. I do like my solution though since I don't have to install a dotnet tool. I'll make sure to do more research on that library at a later date though!

**Comment by Presipope** [comment, marked as answer]
This is what I ended up with. 

I created the following class in my GraphQL project to generate the schema string I would need. 

```c#
    public static class SchemaGenerator
    {
        public static string Generate()
        {
            Schema schema = new Schema()
            {
                Query = new RootQuery(),
                Mutation = new RootMutation()
            };
            schema.RegisterTypeMappings(typeof(RootSchema).Assembly);
            SchemaPrinter printer = new SchemaPrinter(schema);
            string printedSchema = printer.Print();
            return printedSchema;
        }
    }
``` 

Then I created a C# console application in the same solution who's sole purpose was to grab that schema string and write it to the location the client side Relay application needed the schema.graphql file to be. The code ended up looking like this. Probably a little ugly but I ran into some issues running the console application stand alone vs in the build step so I was just working through the directory problems. 

```c#
using MyProject.GraphQL;
using System.Reflection;

string parentDirectoryName = "MyProject.GraphQL.PostBuild";
// where the application is located, usually within the bin folder
string currentDirectoryPath = Assembly.GetExecutingAssembly().Location;
// grab the current parent directory of the executing assemblies location
string parentDirectoryPath = Directory.GetParent(currentDirectoryPath)?.FullName;

// work through all parents until we find the project directory name
while (parentDirectoryPath != null && Path.GetFileName(parentDirectoryPath) != parentDirectoryName)
{
    parentDirectoryPath = Directory.GetParent(parentDirectoryPath)?.FullName;
}
// hopefully we found the projects directory
if (parentDirectoryPath != null)
{
    Console.WriteLine($"Path to {parentDirectoryName} directory: {parentDirectoryPath}");
}
else
{
    Console.WriteLine($"Could not find {parentDirectoryName} directory in {currentDirectoryPath} or any of its parent directories.");
}
// now that we know we are at the project level we can go up one to get the solution directory
string solutionDirectory = Directory.GetParent(parentDirectoryPath)?.FullName;
// constructing the manual path from the solution directory to get to where this file should be saved
string finalDestination = Path.Combine(solutionDirectory, "MyProject.Web", "my-project-client", "src");
// grab the schema defintion 
string schemaString = SchemaGenerator.Generate();
Console.WriteLine($"Writing schema.graphql file to the directory: {finalDestination}");
File.WriteAllText(finalDestination + @"\schema.graphql", schemaString);
``` 

Then inside of the MyProject.GraphQL post build events I added the following command: 

```
$(SolutionDir)MyProject.GraphQL.PostBuild\bin\$(ConfigurationName)\net7.0\MyProject.GraphQL.PostBuild.exe
```
